lib/lists.nix

```diff
@@ -198,8 +198,20 @@ rec {
     default:
     # Input list
     list:
-    let found = filter pred list;
-    in if found == [] then default else head found;
+    let
+      len = length list;
+      recurse = index:
+        let
+          value = elemAt list index;
+        in
+        if index >= len then
+          default
+        else if pred value then
+          value
+        else
+          recurse (index + 1);
+    in
+    recurse 0;
```

Comment:
This will overflow the evaluator stack when the match is sufficiently far into the list.
I think you could do it in O(log n) frames by structuring the control flow as a tree instead of a stack.
The first idea I had for this traversal is something like this:

```
let recurseUp = offset: iterationSize:
       binarySearch in [offset, offset+iterationSize)
       if found then found
       else if out of bounds then throw  # actually binarySearch would throw
       else recurseUp (offset+iterationSize) (iterationSize * 2)
in recurseUp 0 1
```

However, you could implement only a regular binary search, with the downside that the best case isn't as good.

Actually I think Nix should implement a trampoline builtin so we can write tail recursive stuff with some effort. (Or "just" proper tail recursion)

```
# pseudocode
trampoline = v:
  while not v?value
    v = v.thunk
  return v.value
```

... but we don't have that, so the current change is a regression and we'd have to wait a long time or do the binary search thing.
Oh and that trampoline thing is actually a lot like a lazy list if you were to keep the old `v`, but we don't need to get into that. I think?

Reply:
Ugh, good catch! I added a test case for this.

I figured out how to make it work without much complication though, it's pretty fancy, check out the new commit!

Reply:
Ah, `foldl'`. That does iterate, but always a fixed number of times. That makes the solution O(length), so no early cutoff. That could make repeated lookups into the front of a large list significantly slower, but it is already a big improvement over `filter`.

Reply:
Yeah, though iteration should be really fast still, since it will have to evaluate barely anything for every iteration after the match.

Reply:
Yeah. Just wrote #8430, but the attrset allocations required to differentiate between returning and continuing make it seem quite expensive. And anything that avoids using `trampoline` would have to reimplement the idea as part of the caller's logic; no separation of concerns; an overly complicated function.

I think we can take the asymptotic hit for the best cases and optimize the constant factor of the case where the match could be anywhere. I think that's what this achieves now.